### ISSUE TITLE
Comparison of major device number with single LOOPDEV_MAJOR is not indicative of a loop device

### ISSUE THREAD

**Issue by joergschray** [karelzak/util-linux#1202]
It is possible to set up loop devices with major device numbers other than 7:
`dd if=/dev/zero of=/tmp/loop bs=1k count=1; losetup /dev/loop12345678 /tmp/loop`

However such devices can afterwards no longer be manipulated by `losetup`, since the function `is_loopdev` in `lib/loopdev.c` does not recognize them as loop devices. They can neither be listed nor deleted - neither individually nor collectively.

For reference: I am running on kernel release 4.19.158.

**Comment by karelzak** [comment]
Hmm ... frankly not sure how to fix it ;-) 

We can remove the is_loopdev() from some code paths; for example for -d/-D/-a it's probably a good idea, but not sure about other cases.

I see the same problem in Linux kernel where drivers/block/loop.c:is_loop_device() uses the same logic and the function is used for loop configuration and change-fd ioctls. I'll inform kernel devels about it.

**Comment by joergschray** [comment]
The loop device shows up in the `/sys` file system e.g. as `/sys/devices/virtual/block/loop0`. If it is active there is a subdirectory `loop` which contains information about the device e.g. a readable file named `backing_file`. The subdirectory is also present for a device with a differing major number. So the kernel exposes information about it via sysfs correctly.

I have no idea if a utility such as `losetup` should use the `/sys` file system as interface.

**Comment by karelzak** [comment]
I know about it and the losetup already uses /sys.

The problem is how to detect that a random path is a path to the loop device if we do not want to rely on the device name. Anyway, it seems we can use canonicalization of symlink /sys/dev/block/<maj>:<min>/ which points to /sys/devices/virtual/block/loop<n>. I'll implement it.

**Comment by karelzak** [comment]
The problem should be fixed now Thanks for your report!

**Comment by joergschray** [comment]
Thanks for the fix!

**Comment by bebarino** [comment]
This seems to make umount remove the loopback device for me when 'dev' passed into `is_loopdev()` is a loopdev partition and not a device. I think the problem is I have something like the below `lsblk /dev/loop1` and I `umount /dev/loop1p1`. This makes `mnt_context_prepare_umount()` think that it should auto delete the loopback device.

```
NAME       MAJ:MIN RM  SIZE RO TYPE MOUNTPOINT
loop1        7:1    0  6.5G  0 loop 
 ├─loop1p1  259:28   0    4G  0 part 
 ├─loop1p2  259:29   0   32M  0 part 
 ├─loop1p3  259:30   0  2.3G  0 part 
 ├─loop1p4  259:31   0   32M  0 part 
 ├─loop1p5  259:32   0    2M  0 part 
 ├─loop1p6  259:33   0  512B  0 part 
 ├─loop1p7  259:35   0  512B  0 part 
 ├─loop1p8  259:36   0   16M  0 part 
 ├─loop1p9  259:37   0  512B  0 part 
 ├─loop1p10 259:38   0  512B  0 part 
 ├─loop1p11 259:39   0  512B  0 part 
 └─loop1p12 259:40   0   64M  0 part 
```

I haven't fully traced the code but reverting this patch makes the loop device stick around after unmounting the partitions. Is the logic correct? I think one option would be to look in sysfs and see if the path has a 'partition' file with a value of 1 in it. If it does, this isn't a loopback device, but instead a loopback partition. I just stripped the parent directory off the path and checked if it started with 'loop' again and that also seemed to work. Hacky patch attached. I suppose we could also look for 'block' instead of not 'loop'. Maybe even better would be to check if `/sys/block/<p>` points to the same place as `cn` and then we know it's a block device instead of a partition?

```
diff --git a/lib/loopdev.c b/lib/loopdev.c
index 7f299ebf3..f8b154b79 100644
--- a/lib/loopdev.c
+++ b/lib/loopdev.c
@@ -653,6 +653,8 @@ int is_loopdev(const char *device)
                if (cn)
                        p = stripoff_last_component(cn);
                rc = p && startswith(p, "loop");
+               p = stripoff_last_component(cn);
+               rc = p && !startswith(p, "loop");
                free(cn);
        }
 
```

In sysfs the partition is `/sys/devices/virtual/block/loop1/loop1p1`

```
$ ls /sys/devices/virtual/block/loop1/partition
ls: cannot access '/sys/devices/virtual/block/loop1/partition': No such file or directory
$ cat /sys/devices/virtual/block/loop1/loop1p1/partition 
1
```

**Comment by karelzak** [comment]
@bebarino thanks for your report! This is a regression (unfortunately released in v2.36.2).

It's better to not depend on the device name, so I have used partition/whole-disk detection in sysfs for this purpose.

**Comment by bebarino** [comment]
Thanks the fix works for me.
